lib/rubygems/commands/build_command.rb

```diff
@@ -57,13 +57,30 @@ def usage # :nodoc:
   end
 
   def execute
-    gemspec = get_one_gem_name
+    build_gem(gem_name)
+  end
+
+  private
+
+  def gem_name
+    get_one_optional_argument || find_gemspec
+  end
+
+  def find_gemspec
+    gemspecs = Dir.glob("*.gemspec").sort
 
-    unless File.exist? gemspec
-      gemspec += '.gemspec' if File.exist? gemspec + '.gemspec'
+    if gemspecs.size > 1
+      alert_error "Multiple gemspecs found: #{gemspecs}, please specify one"
+      terminate_interaction(1)
     end
 
-    if File.exist? gemspec
+    gemspecs.first
+  end
+
+  def build_gem(gem_name = get_one_optional_argument)
```

Comment:
This method is never called without an argument, so the optional value is unnecessary. Also, the parameter (`gem_name`) is called with the same name as a method, that confuses me a bit, personally.

I would remove the `build_gem` and `gem_name` methods and inline the whole logic inside the `execute` method, because the execute method now just delegates to `build_gem`, which seems like an unnecessary indirection to me that doesn't help readability.

Something like

```ruby
def execute
  gem_name = get_one_optional_argument || find_gemspec
  gemspec = File.exist?(gem_name) ? gem_name : "#{gem_name}.gemspec"

  # ... and the rest of the current `build_gem` logic ...
end
```

Reply:
I removed the `gem_name` method.

But i do prefer:

```ruby
def execute
  gem_name = get_one_optional_argument || find_gemspec
  build_gem(gem_name)
end
```

over 

```ruby
def execute
  gem_name = get_one_optional_argument || find_gemspec
  gemspec = File.exist?(gem_name) ? gem_name : "#{gem_name}.gemspec"

    if File.exist?(gemspec)
      spec = Gem::Specification.load(gemspec)

      if options[:build_path]
        Dir.chdir(File.dirname(gemspec)) do
          spec = Gem::Specification.load(File.basename(gemspec))
          build_package(spec)
        end
      else
        build_package(spec)
      end

    else
      alert_error "Gemspec file not found: #{gemspec}"
      terminate_interaction(1)
    end
end
```
In the former I don't have a doubt what `execute` is exactly doing